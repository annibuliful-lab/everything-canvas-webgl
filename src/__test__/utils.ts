export function createCanvas() {
  const canvas = document.createElement("canvas");
  const ctx = canvas.getContext("2d");
  return { canvas, ctx };
}
