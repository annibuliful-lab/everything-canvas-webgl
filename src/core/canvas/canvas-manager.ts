import { I2dPosition } from "../@types/Object";
import { ShapeBaseObject } from "../shapes/object";

export type Dimension = {
  width: number;
  height: number;
};

export interface ICanvasObject extends ShapeBaseObject {
  draw(ctx: CanvasRenderingContext2D): void;
  contains(position: I2dPosition): boolean;
  setCoords(): void;
  get boundingBox(): {
    topLeft: I2dPosition;
    topRight: I2dPosition;
    bottomLeft: I2dPosition;
    bottomRight: I2dPosition;
  };

  get left(): number;
  get right(): number;
  get top(): number;
  get topLeft(): I2dPosition;
  get topRight(): I2dPosition;
  get bottom(): number;
  get bottomLeft(): I2dPosition;
  get bottomRight(): I2dPosition;

  get centerPoint(): {
    top: number;
    left: number;
  };

  toObject(): any;
}

export class CanvasManager {
  element: HTMLCanvasElement;
  context: CanvasRenderingContext2D;
  width: number = 0;
  height: number = 0;
  private _objects: ICanvasObject[] = [];

  constructor(canvas: HTMLCanvasElement) {
    this.element = canvas;
    this.context = canvas.getContext("2d") as CanvasRenderingContext2D;
  }

  add<T extends ICanvasObject>(object: T) {
    this._objects.push(object);
  }

  remove(id: string) {
    this._objects = this._objects.filter((object) => object.id !== id);
  }

  getObjects() {
    return this._objects;
  }

  clear() {
    this.context.clearRect(0, 0, this.width, this.height);
  }

  render() {
    // Need to clear in every call render
    this.clear();

    for (const object of this._objects) {
      object.draw(this.context);
    }
  }

  setDimension(dimension: Dimension) {
    const { width, height } = dimension;
    this.element.width = width;
    this.element.height = height;
    this.width = width;
    this.height = height;
  }
}
