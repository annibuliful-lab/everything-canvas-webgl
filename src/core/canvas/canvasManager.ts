import { ShapeBaseObject } from "../shapes/object";

export type Dimension = {
  width: number;
  height: number;
};

export interface ICanvasObject extends ShapeBaseObject {
  draw(ctx: CanvasRenderingContext2D): void;
  centerPoint: {
    top: number;
    left: number;
  };
}

export class CanvasManager {
  element: HTMLCanvasElement;
  context: CanvasRenderingContext2D;
  width: number = 0;
  height: number = 0;
  private _objects: ICanvasObject[] = [];

  constructor(canvas: HTMLCanvasElement) {
    this.element = canvas;
    this.context = canvas.getContext("2d") as CanvasRenderingContext2D;
  }

  add<T extends ICanvasObject>(object: T) {
    this._objects.push(object);
  }

  remove(id: string) {
    this._objects = this._objects.filter((object) => object.id !== id);
  }

  getObjects() {
    return this._objects;
  }

  clear() {
    this.context.clearRect(0, 0, this.width, this.height);
  }

  render() {
    // Need to clear in every call render
    this.clear();

    for (const object of this._objects) {
      object.draw(this.context);
    }
  }

  setDimension(dimension: Dimension) {
    const { width, height } = dimension;
    this.element.width = width;
    this.element.height = height;
    this.width = width;
    this.height = height;
  }
}
