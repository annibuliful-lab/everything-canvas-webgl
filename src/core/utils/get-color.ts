import { colorNameToHex } from "./color-name-to-hex";
import { convertHexToRGBA } from "./hex-to-rgba";

export function getColor(color: string, opacity = 1) {
  return color.startsWith("#")
    ? convertHexToRGBA(color, opacity)
    : convertHexToRGBA(colorNameToHex(color), opacity);
}
