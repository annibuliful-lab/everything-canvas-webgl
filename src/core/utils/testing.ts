import { ObjectPositioDimension } from "../@types/Object";
import { Circle } from "../shapes/circle";
import { Rectangle } from "../shapes/rectangle";

const BASE_PARAM = {
  x: 10,
  y: 12,
  radius: 100,
  id: "BASE_PARAM",
};

export function createRect(data?: ObjectPositioDimension & { id: string }) {
  return new Rectangle(data ?? BASE_PARAM);
}

const BASE_PARAM_CIRCLE = {
  x: 10,
  y: 12,
  radius: 100,
  id: "BASE_PARAM",
};

export function createCircle(
  data?: ObjectPositioDimension & { id: string; radius: number }
) {
  return new Circle(data ?? BASE_PARAM_CIRCLE);
}
