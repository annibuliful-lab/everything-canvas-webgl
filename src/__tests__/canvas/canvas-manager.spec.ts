import { CanvasManager } from "../../core/canvas/canvas-manager";
import "jest-canvas-mock";
import { createCircle, createRect } from "../../core/utils/testing";

describe("CanvasManager class", () => {
  let canvasManager: CanvasManager;
  beforeEach(() => {
    const canvas = document.createElement("canvas");
    canvasManager = new CanvasManager(canvas);
  });

  it("it should be defined", () => {
    const canvas = document.createElement("canvas");
    expect(new CanvasManager(canvas)).toBeTruthy();
  });

  describe("add function", () => {
    it("it should add rectangle", () => {
      const rect = createRect();
      canvasManager.add(rect);
      expect(
        canvasManager.getObjects().map((o) => o.toObject())
      ).toMatchSnapshot();
    });

    it("it should add circle", () => {
      const circle = createCircle();
      canvasManager.add(circle);
      expect(
        canvasManager.getObjects().map((o) => o.toObject())
      ).toMatchSnapshot();
    });

    it("it should add multiple objects", () => {
      const rect = createRect();
      const circle = createCircle();

      canvasManager.add(rect);
      canvasManager.add(circle);
      expect(
        canvasManager.getObjects().map((o) => o.toObject())
      ).toMatchSnapshot();
    });
  });
});
