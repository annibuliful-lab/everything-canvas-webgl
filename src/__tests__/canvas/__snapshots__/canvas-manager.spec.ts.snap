// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanvasManager class add function it should add circle 1`] = `
Array [
  Object {
    "border": Object {
      "borderColor": "rgba(0,0,0,0)",
      "borderWidth": 0,
    },
    "boundingBox": Object {
      "bottomLeft": Object {
        "x": -90,
        "y": 12,
      },
      "bottomRight": Object {
        "x": 110,
        "y": 12,
      },
      "topLeft": Object {
        "x": -90,
        "y": 112,
      },
      "topRight": Object {
        "x": 110,
        "y": 112,
      },
    },
    "fill": "#000000",
    "height": 0,
    "opacity": 1,
    "radius": 100,
    "scale": Object {
      "scaleX": 1,
      "scaleY": 1,
    },
    "scaleX": 1,
    "scaleY": 1,
    "width": 0,
    "x": 10,
    "y": 12,
  },
]
`;

exports[`CanvasManager class add function it should add multiple objects 1`] = `
Array [
  Object {
    "border": Object {
      "borderColor": "rgba(0,0,0,0)",
      "borderWidth": 0,
    },
    "boundingBox": Object {
      "bottomLeft": Object {
        "x": 10,
        "y": 12,
      },
      "bottomRight": Object {
        "x": 10,
        "y": 12,
      },
      "topLeft": Object {
        "x": 10,
        "y": 12,
      },
      "topRight": Object {
        "x": 10,
        "y": 12,
      },
    },
    "centerPoint": Object {
      "left": 5,
      "top": 6,
    },
    "fill": "#000000",
    "height": 0,
    "opacity": 1,
    "scale": Object {
      "scaleX": 1,
      "scaleY": 1,
    },
    "scaleX": 1,
    "scaleY": 1,
    "width": 0,
    "x": 10,
    "y": 12,
  },
  Object {
    "border": Object {
      "borderColor": "rgba(0,0,0,0)",
      "borderWidth": 0,
    },
    "boundingBox": Object {
      "bottomLeft": Object {
        "x": -90,
        "y": 12,
      },
      "bottomRight": Object {
        "x": 110,
        "y": 12,
      },
      "topLeft": Object {
        "x": -90,
        "y": 112,
      },
      "topRight": Object {
        "x": 110,
        "y": 112,
      },
    },
    "fill": "#000000",
    "height": 0,
    "opacity": 1,
    "radius": 100,
    "scale": Object {
      "scaleX": 1,
      "scaleY": 1,
    },
    "scaleX": 1,
    "scaleY": 1,
    "width": 0,
    "x": 10,
    "y": 12,
  },
]
`;

exports[`CanvasManager class add function it should add rectangle 1`] = `
Array [
  Object {
    "border": Object {
      "borderColor": "rgba(0,0,0,0)",
      "borderWidth": 0,
    },
    "boundingBox": Object {
      "bottomLeft": Object {
        "x": 10,
        "y": 12,
      },
      "bottomRight": Object {
        "x": 10,
        "y": 12,
      },
      "topLeft": Object {
        "x": 10,
        "y": 12,
      },
      "topRight": Object {
        "x": 10,
        "y": 12,
      },
    },
    "centerPoint": Object {
      "left": 5,
      "top": 6,
    },
    "fill": "#000000",
    "height": 0,
    "opacity": 1,
    "scale": Object {
      "scaleX": 1,
      "scaleY": 1,
    },
    "scaleX": 1,
    "scaleY": 1,
    "width": 0,
    "x": 10,
    "y": 12,
  },
]
`;

exports[`CanvasManager class remove function it should remove rectangle 1`] = `
Array [
  Object {
    "border": Object {
      "borderColor": "rgba(0,0,0,0)",
      "borderWidth": 0,
    },
    "boundingBox": Object {
      "bottomLeft": Object {
        "x": -90,
        "y": 12,
      },
      "bottomRight": Object {
        "x": 110,
        "y": 12,
      },
      "topLeft": Object {
        "x": -90,
        "y": 112,
      },
      "topRight": Object {
        "x": 110,
        "y": 112,
      },
    },
    "fill": "#000000",
    "height": 0,
    "opacity": 1,
    "radius": 100,
    "scale": Object {
      "scaleX": 1,
      "scaleY": 1,
    },
    "scaleX": 1,
    "scaleY": 1,
    "width": 0,
    "x": 10,
    "y": 12,
  },
]
`;
